Data/ByteString/Streaming/Char8.hs

```diff
@@ -587,6 +588,67 @@ unwords = intercalate (singleton ' ')
 {-# INLINE unwords #-}
 
 
+{- | 'lineSplit' turns a ByteString into a connected stream of ByteStrings at
+     divide after a fixed number of newline characters. 
+     Unlike most of the string splitting functions in this library, 
+     this function preserves newlines characters. 
+
+     Like 'lines', this function properly handles both @\\n@ and @\\r\\n@
+     endings regardless of the current platform.
+     
+     >>> let planets = ["Mercury","Venus","Earth","Mars","Saturn","Jupiter","Neptune","Uranus"]
+     >>> S.mapsM_ (\x -> putStrLn "Chunk" >> Q.putStrLn x) $ Q.lineSplit 3 $ Q.string $ L.unlines planets
+     Chunk
+     Mercury
+     Venus
+     Earth
+
+     Chunk
+     Mars
+     Saturn
+     Jupiter
+
+     Chunk
+     Neptune
+     Uranus
+
+-}
+lineSplit :: forall m r. Monad m 
+  => Int -- ^ number of lines per group
+  -> ByteString m r -- ^ stream of bytes
+  -> Stream (ByteString m) m r
+lineSplit !n0 text0 = loop1 0 text0
+  where
+    n :: Int
+    !n = max n0 1
+    loop1 :: Int -> ByteString m r -> Stream (ByteString m) m r
+    loop1 !counter text =
+      case text of
+        Empty r -> Return r
+        Go m -> Effect $ liftM (loop1 counter) m
+        Chunk c cs
+          | B.null c -> loop1 counter cs
+          | otherwise -> Step (loop2 counter text)
+    loop2 :: Int -> ByteString m r -> ByteString m (Stream (ByteString m) m r)
+    loop2 !counter text =
+      case text of
+        Empty r -> Empty (Return r)
+        Go m -> Go $ liftM (loop2 counter) m
+        Chunk c cs ->
+          let !numNewlines = B.count 10 c
```

Comment:
What is the purpose of the magic number `10`? (repeated below as well)

Reply:
In ASCII encodings and in UTF-8, a byte with decimal value 10 is a newline character:

>>> Data.Char.ord '\n'
10

Reply:
Maybe define that in a variable (since you use `10` twice)?

I wonder if it's likely/possible for there to be a carriage return in there that you would have to consider...

Reply:
I would be fine with defining

    newline :: Word8
    newline = 10

since GHC is just going to inline it anyway. The carriage return is a situation that I considered. Fortunately, since carriage returns show up **before** newlines in the dos `\r\n` line-end encoding, this works fine for them to.

Reply:
Please do (even if it's just within a `where` clause)!

I can never remember whether GHC converts `\r\n` to just `\n` or not, and in the case of reading a raw file in terms of bytes I suspect it won't.  In that case, I think having `\r\n` might break `null` check you have there.

(In practical terms `\r` and `\n\r` are probably so unlikely nowadays that `\r\n` is the only corner case that needs to be considered here.)

Reply:
I see that `lines` hard-codes `10` and `13` (`\r`) as well... but it seems to support splitting on `\r\n` by checking the character before the `B.findIndices`.

Reply:
`lines` attempts to support splitting on `\r\n`, but [it's actually broken](https://github.com/michaelt/streaming-bytestring/issues/9).